### PR TITLE
[mlir] Add `convertInstruction` and `getSupportedInstructions` to `LLVMImportInterface`

### PR DESCRIPTION
**Pull request by fabianmcg**
This patch adds the `convertInstruction` and `getSupportedInstructions` to `LLVMImportInterface`, allowing any non-LLVM dialect to specify how to import LLVM IR instructions.

This patch is necessary for https://github.com/llvm/llvm-project/pull/73057

### CODE REVIEW

File: mlir/test/lib/Dialect/Test/TestFromLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Support/SourceMgr.h"
+
+using namespace mlir;
+using namespace test;
+
+namespace {
+inline ArrayRef<unsigned> getSupportedInstructionsImpl() {
```
Comment (Dinistro) [inline mlir/test/lib/Dialect/Test/TestFromLLVMIRTranslation.cpp:34]: Nit: It's preferred to use static functions over wrapping them into anonymous namespaces. The reasoning is that the internal nature of the function is directly visible from its signature.

File: mlir/test/Target/LLVMIR/Import/test.ll [hunk 2]
```diff
@@ -0,0 +1,11 @@
+; RUN: mlir-translate -test-import-llvmir %s | FileCheck %s
+
+; CHECK-LABEL: @custom_load
+; CHECK-SAME:  %[[PTR:[a-zA-Z0-9]+]]
```
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/test.ll:4]:
```suggestion
; CHECK-SAME:  %[[PTR:[:alnum:]+]]
```
This would be a bit simpler, I think.

**Review by Dinistro — COMMENTED** [review]
Thanks for this extension. This will make a lot of sense for you pointer dialect change, and will allow us to gradually transition into using it.

Once the transition period is over, we should consider to also add the same hook to the LLVM dialect import interface.

**Review by Dinistro — APPROVED** [review]
LGTM, but please wait a bit for @gysit to take a look as well.

**Review by gysit — APPROVED** [review]
Nice! 

LGTM modulo one nit.

File: mlir/include/mlir/Target/LLVMIR/LLVMImportInterface.h [hunk 3]
```diff
@@ -100,9 +114,27 @@ class LLVMImportInterface
...
+            return instructionToDialect.count(id);
+          });
+      if (instIt != iface.getSupportedInstructions().end()) {
+        return emitError(
+            UnknownLoc::get(iface.getContext()),
+            llvm::formatv(
+                "expected unique conversion for instruction ({0}), but "
+                "got conflicting {1} and {2} conversions",
+                *it, iface.getDialect()->getNamespace(),
+                instructionToDialect.lookup(*it)
```
[the remark this answers is not in the document]
Comment (fabianmcg) [reply mlir/include/mlir/Target/LLVMIR/LLVMImportInterface.h]: I changed them to `intrinsicIt` and `instructionIt`, as `instIt` was already in use.

File: mlir/include/mlir/Target/LLVMIR/LLVMImportInterface.h [hunk 4]
```diff
@@ -66,6 +75,11 @@ class LLVMImportDialectInterface
   /// returns the list of supported intrinsic identifiers.
   virtual ArrayRef<unsigned> getSupportedIntrinsics() const { return {}; }
 
+  /// Hook for derived dialect interfaces to publish the supported instructions.
+  /// As every LLVM IR instructions has a unique integer identifier, the
+  /// function returns the list of supported instructions identifiers.
+  virtual ArrayRef<unsigned> getSupportedInstructions() const { return {}; }
```
Comment (joker-eph) [inline mlir/include/mlir/Target/LLVMIR/LLVMImportInterface.h:81]: What if multiple dialect return overlapping results here?

Also the documentation is saying what this does, but not why it is useful or how this should be used, can you expand?

**Comment by Dinistro** [comment]
Can we land this?

**Comment by fabianmcg** [comment]
@gysit, @zero9178  see @Dinistro comment on landing?

**Comment by gysit** [comment]
Yes this is good to go! 

Thanks for adding this additional extensibility to the import.

**Review by zero9178 — APPROVED** [review]
Feel free to land!

**Pull request merged**